### DISCUSSION TITLE
Как сделать доп цикл, к циклу.

### DISCUSSION THREAD

**Discussion by Qb1cc** [totumonline/totum-ru-issues-and-discussions#88] [Как это сделать?] [answered]
При создании заказа в цикле, иногда нужно сделать дополнительный заказ. 
В старой заказной это реализовано с помощью дробей. Допустим был заказ №1, мы делаем дубль( и меняем всё нам нужно ) с номером заказа №1/1
И далее можно уже чисто по номеру 1, увидеть еще и дробные заказы к основному. 

Как реализовать подобное в Тотум? Вроде что то кнопки, типо сделать дубль цикла текущего id1, но при этом в в таблице циклов в колонке где болтаются номера заказов с форматированием писать №1/1, и потом фильтром по этой колонке отфильтровать уже

**Comment by Qb1cc** [comment]
Я могу вкинуть заказную для автозапчастей + видеоинструкцию к ней. Вы поймёте все механизмы сразу же . Программа десятилетия отработала, но в новых условиях нужно прикручивать оповещения(что успешно реализовано на Тотум)  со своего шлюза с симкой на йоте), а исходников уже нет, как и разрабов.

( но я не могу на опенсоурс это выложить, только в лс для ознакомления )

**Comment by ttm-alexeypolunin** [comment]
Я напишу по возможности завтра — у нас в двух проектах была такая задача.

**Comment by ttm-alexeypolunin** [comment, marked as answer]
### [Видео на YouTube —>](https://youtu.be/jubNcx4OngY)

Коды полей:

**duplicate:**
```
= : $#duplicatedId
```

**first_number:**
```
= : if(condition: #duplicate = 0; then: $calc; else: $sel_first)

calc: $sel + 1

	sel: select(table: $#ntn; field: 'first_number'; where: 'duplicate' = 0; order: 'id' desc)

sel_first: select(table: $#ntn; field: 'first_number'; where: 'id' = #duplicate)
```

**second_number:**
```
= : if(condition: #duplicate != 0; then: $calc; else: "")

calc: $sel + 1

	sel: select(table: $#ntn; field: 'second_number'; where: 'first_number' = #first_number; order: 'id' desc)
```

**number:**
```
= : if(condition: #second_number != ""; then: str`#first_number + " / " + #second_number`; else: #first_number)
```

**Reply by Qb1cc** [reply]
Огромное спасибо за объяснение.

**Comment by Qb1cc** [comment]
Получил сразу две поблемы после добавления. 
Первая это вывод в тайтл ID цикла, но нужно выводить номер теперь из колонки
![image](https://user-images.githubusercontent.com/84081305/157410996-340ad90e-474b-4dfd-8722-ac5e2f8dcb47.png)

Вторая проблема, это некорректное оприходование. Там примерно тоже самое, в Uniq уходит id заказа, но нужно передать в него номер заказа из таблицы циклов с тех поля number из этого урока.


![image](https://user-images.githubusercontent.com/84081305/157411579-ed5d2bd6-a495-40dd-83d8-ebec017795f8.png)
![image](https://user-images.githubusercontent.com/84081305/157411644-35cc72c3-1eaf-4011-ab3c-422ac124571c.png)

**Reply by ttm-alexeypolunin** [reply]
Не привязывайтесь к номеру — привязывайтесь к `id` — номер только для людей. Это вообще общий подход.

Дл тайтла перенастройте поле из которого берется информация для `str` в коде форматирования таблицы.

Поставьте номер как Основное поле в натсройках таблицы циклов — тогда в дереве при открытии цикла тоже будет отображаться ок.

**Reply by Qb1cc** [reply]
Да я разобрался как работает пример выше. Большое спасибо, но у меня несколько другая проблема возникла. 

Что бы в цикл получить значение(отмеченное стрелкой)  из таблицы циклов
![image](https://user-images.githubusercontent.com/84081305/157610882-2bf8c625-da13-44bf-9b45-1844f379ef1c.png)

, нужно каким то необычным образом сделать сортировку не по ID потому что ID и поле номер заказа не одинаковые и на выходе мы получаем не совсем корректное оприходование товара из урока 

[totumonline/totum-ru-issues-and-discussions/discussions/75]

![image](https://user-images.githubusercontent.com/84081305/157608702-db6a091e-ae93-4295-a70d-151d3fe1e6c0.png)
![image](https://user-images.githubusercontent.com/84081305/157609288-7ec1e4cb-f345-4a7b-a703-fceb1715f9a2.png)

Тоже самое и для тайтла, для того что б в тайтл падал номер заказа , а не просто ID цикла, нужно какое то поле сделать в самом цикле куда будет падать форматированный номер заказа примером выше.

**Reply by ttm-alexeypolunin** [reply]
Сортировка — посмотрите функцию [setRowsOrder](https://docs.totum.online/functions#fn-setRowsOrder) и поставьте поле номер как поле сортировки в настройках таблицы и еще можно в h_insert сортировать как вам нужно.

Перенести номер заказа в поле внутри цикла — 

1. Создаем поле внутри цикла

2. Пишем в нем Код

```
=: select(table: 'cycles_table'; field: 'field_with_number'; where: 'id' = $#nci)

// номер цикла это id в таблице циклов и мы можем любыми функциями изнутри цикла обращаться
```

4. Пересчитайте циклы после добавления поля

**Reply by Qb1cc** [reply]
Всё получилось. Оказалось я  затупил с select и не правильно делал where. Теперь всё работает как нужно.

**Comment by Alexmeb** [comment]
А если усложнить немного задачу. Есть номер 1700 например. С дублированием получается 1700.1
Нужно чтобы при дублировании, следующий становился 1700.2, а родитель 1700.1
Т.е. добавляем заказ, он имеет номер 1700. Как только начинаем дублировать, идут .2  .3  .4 а первый меняется на .1

**Reply by ttm-alexeypolunin** [reply]
В `second_number` написать `код-действия при добавлении`, который будет делать `set` в строку с самым первым донором.
